src/tls/test-server.c

```diff
@@ -194,51 +196,47 @@ assert_https_outcome (TestCase *tc,
       g_assert_cmpint (gnutls_credentials_set (session, GNUTLS_CRD_CERTIFICATE, xcred), ==, GNUTLS_E_SUCCESS);
 
       ret = gnutls_handshake (session);
-      if (expected_handshake_result == GNUTLS_E_SUCCESS)
+      if (ret != GNUTLS_E_SUCCESS)
         {
-          const gnutls_datum_t *server_certs;
-          unsigned server_certs_len;
-          ssize_t len;
-
-          if (ret != GNUTLS_E_SUCCESS)
+          if (expect_tls_failure)
+            exit (0);
+          else
             g_error ("Handshake failed: %s", gnutls_strerror (ret));
+        }
 
-          /* check server certificate */
-          server_certs = gnutls_certificate_get_peers (session, &server_certs_len);
-          g_assert (server_certs);
-          g_assert_cmpuint (server_certs_len, ==, expected_server_certs);
-
-          /* send request, read response */
-          g_assert_cmpint (gnutls_record_send (session, request, sizeof (request)), ==, sizeof (request));
-          len = gnutls_record_recv (session, buf, sizeof (buf) - 1);
-          if (expected_read_error != 0)
-            {
-              g_assert_cmpint (len, ==, expected_read_error);
-              close (fd);
-              exit (0);
-            }
-
-          g_assert_cmpint (len, >=, 100);
-          buf[len] = '\0'; /* so that we can use string functions on it */
-          /* This succeeds (200 OK) when building in-tree, but fails with dist-check due to missing doc root */
-          if (strstr (buf, "200 OK"))
-            {
-              cockpit_assert_strmatch (buf, "HTTP/1.1 200 OK\r\n"
-                                            "Content-Type: text/html\r\n"
-                                            "Content-Security-Policy: connect-src 'self' https://localhost wss://localhost;*");
-            }
-          else
-            {
-              cockpit_assert_strmatch (buf, "HTTP/1.1 404 Not Found\r\nContent-Type: text/html*");
-            }
+      /* check server certificate */
+      server_certs = gnutls_certificate_get_peers (session, &server_certs_len);
+      g_assert (server_certs);
+      g_assert_cmpuint (server_certs_len, ==, expected_server_certs);
 
-          g_assert_cmpint (gnutls_bye (session, GNUTLS_SHUT_RDWR), ==, GNUTLS_E_SUCCESS);
+      /* send request, read response */
+      len = gnutls_record_send (session, request, sizeof (request));
+      if (len < 0 && expect_tls_failure)
+        exit (0);
+      g_assert_cmpint (len, ==, sizeof (request));
+
+      len = gnutls_record_recv (session, buf, sizeof (buf) - 1);
+      if (len < 0 && expect_tls_failure)
+        exit (0);
+      g_assert_cmpint (len, >=, 100);
+
```

Comment:
can you also assert that len<= sizeof buf - 1 before writing this byte here?

Reply:
nicely spotted, thanks!